qa/full-cluster-restart/build.gradle

```diff
@@ -54,11 +54,20 @@ for (Version bwcVersion : bwcVersions.indexCompatible) {
     systemProperty 'tests.is_old_cluster', 'true'
   }
 
+  TaskProvider upgradeCluster = tasks.register("${baseName}#upgradeCluster") {
```

Comment:
The tests in the old cluster also create state that is verified in the upgrade. 
I'm not sure creating a task for the upgrade is worthwhile as we won't be able to run the tests on the upgrade regardless, so  the check in `doFirst` should be sufficient.
We made it a goal not to have tasks specifically deal with clusters so this feels like a step backward. 

FYI, I have plans to work on a bwc plugin to avoid some of the repetition in these tests.
The way I see it it should build on a base plugin that implements the concept of a test chain that needs to run in an all or nothing state, with dependencies correctly configured.

Reply:
Sorry, I wanted to explain this in the PR description but didn't get a chance to update it.

The reason for the separate upgrade tasks is for input tracking. We can't do the upgrade in a `doFirst()` because the upgrade changes the state of the node, which is an input. We need to track what distributions the cluster is currently using as part of the input to the task. Otherwise, if we had three identical test tasks, they'd all be cached after the first ran because no inputs would have changed. To track the "upgrade" as a input, it has to happen before the task executes, so it has to either happen in a separate task. It can't be in the previous task's `doLast`, again, because if that task gets cached, then the upgrade won't happen, which throws off the inputs.

Reply:
I 100% agree we should make this whole concept more first-class somehow, and have a plugin setup these tasks for us, but I think we can do that as part of designing a holistic testing DSL. We need to consider this use case as part of that.

Reply:
@mark-vieira I think it could go in the previous tasks's `doLast` rather than adding more tasks. It's true that it won't run if that is cached, but in that case the tests will fail regardless.  The reality is that the tests we run before the upgrade alter the state of the cluster and subsequent tests depend on that state, so to correctly track things we would have to have some for of dump ( indices, mappings, etc, etc ) from the cluster tracked as input. That's more than it's worth, so  I don't think we should deal with the fact that some task could be skipped, if that happens the next  one will fail regardless, ideally they should all be skipped or none of them should 

That being said, we know  all the versions a cluster will use up-front.
I was under the impression that we do consider all versions as input, at least that was my intent when I made that change, if it's not the case perhaps we should fix that. That means that the only difference in terms of input will be the system property that tells the tests what stage the cluster is at.

Reply:
> That being said, we know all the versions a cluster will use up-front.

We cannot use all versions as the input. We have to use _the_ version used by the cluster at that time as the input. Otherwise, if I have three test task, they all have identical inputs and will use each others' cache artifacts, which  is wrong.

> That means that the only difference in terms of input will be the system property that tells the tests what stage the cluster is at.

This is effectively what `currentDistro` is. And this _has_ to be set to the correct value whether the task will be cached or not so that the inputs are always the same. This is why we can't set this in a `doLast` because it must _always_ happen.

There is no way to avoid this situation with mutable nodes. If each test task used a unique immutable cluster config we wouldn't have to do this, but when we mutate stuff that affect test execution, we have to do that _outside_ of the task which tracks that as an input. There's simply no reliable workaround. The only downside of this is just additional tasks. It's not really that big of a deal, especially since we intend to eventually bake this boilerplate into a plugin.

Reply:
Each  test has a system property specific to the task right now, which would prevent the caching behavior you describe, so we would not need the additional tasks, sure the upgrade wouldn't happen, but since the previous task was skipped the current tests will fail regardless.
Granted the only problem is that this is not generic, as in it replies on the tests setting up such a property to avoid caching the results incorrectly, but this is something we know we want to improve on.  

I do think the additional tasks make things harder to understand and I was very deliberate in avoiding them in the testclusters implementation.

Reply:
Ah, I see this now. I'm not a big fan on relying on some tangential input here but that's probably fine until we refactor this later. I'll move the upgrade to the prior task's `doLast()` and rip out these extra tasks.